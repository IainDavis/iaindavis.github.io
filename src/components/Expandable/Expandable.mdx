import { Meta, Canvas } from '@storybook/addon-docs';
import * as ExpandableStories from './Expandable.stories';

<Meta of={ExpandableStories} />

# Expandable
The 'Expandable' component wraps a block of content, making it
selectively shown (expanded) or hidden (collapsed). When expanded, it
provides a visual indication of the extent of the expandable block.

<details>
<summary>Rationale/Use Case</summary>

>    The Docusaurus default styling for the &lt;details&gt; block uses the
>    same styling as admonitions and alerts, _i.e._ a bright, colored
>    border which separates the content from the rest of the text.
>
>    The **Expandable** component is for content that you want to
>    present to the user in an opt-in (or opt-out) manner, without also
>    removing it from the normal flow of content or unduly highlighting it.
</details>

<br/>

## Parameters
```typescript
type ExpandablePropsType = {
    startExpanded?: boolean;
    expandPrompt?: string;
    collapsePrompt?: string;
}
```

---

## Behavior
### Default
**Expandable** renders the expandable content in an opt-in manner, by default
<Canvas of={ExpandableStories.Default} />

### Pre-Expanded
Supply the 'startExpanded' attribute for 'opt-out' behavior
<Canvas of={ExpandableStories.PreExpanded} />

### With Custom Expand/Collapse Prompts
The text prompts for the Expand/Collapse prompts are customizable
<Canvas of={ExpandableStories.CustomExpandCollapsePrompt} />

### Nested
An **Expandable** block can be nested within another **Expandable** block
<Canvas of={ExpandableStories.Nested} />