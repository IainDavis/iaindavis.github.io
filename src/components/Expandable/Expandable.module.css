html {
    --secondary-light: lightGray;
    --secondary-dark: gray;
}

[data-theme='dark'] {
    --secondary-light: gray;
    --secondary-dark: lightGray;
}

.container {
    padding-bottom: 1em;
}

.toggle {
    border-width: 0px;
    background-color: inherit;
    color: var(--secondary-dark);
    font-size: smaller;
    margin-left: 1em;
    margin-top: 1em;
    line-height: 50%;
}

.expandableBlock {
    position: relative;
    margin-left: 1em;
    padding-left: 1.3em;
    border-left: 1px solid;
    border-color: var(--secondary-light)
}

.expandableBlock:before {
    content: " ";
    background: var(--secondary-light);
    position: absolute;
    left: -0.125em;
    width: 0.25em;
    height: 0.25em;
    border-radius: 50%;
}

.expandableBlock:after {
    content: " ";
    background: var(--secondary-light);
    position: absolute;
    left: -0.125em;
    width: 0.25em;
    height: 0.25em;
    bottom: -0.2em;
    border-radius: 50%;
}