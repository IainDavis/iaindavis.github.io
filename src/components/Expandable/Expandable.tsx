import React, { PropsWithChildren,  useEffect,  useState } from 'react';
import styles from './Expandable.module.css';
import getString from '../../strings/getString';

export type ExpandablePropsType = {
    startExpanded?: boolean;
    expandPrompt?: string;
    collapsePrompt?: string;
}

const Expandable: React.FC<PropsWithChildren<ExpandablePropsType>> = ({
    startExpanded = false,
    expandPrompt,
    collapsePrompt,
    children
}) => {
    const effectiveExpandPrompt = expandPrompt || getString('expandable-component-default-expand-prompt');
    const effectiveCollapsePrompt = collapsePrompt || getString('expandable-component-default-collapse-prompt');

    const [isExpanded, setExpanded] = useState(startExpanded);

    useEffect(() => {
        setExpanded(() => startExpanded);
    }, [startExpanded])

    const handleClickToggle = () => {
        setExpanded(currentlyExpanded => !currentlyExpanded);
    }

    return (
        <div className={styles.container}>
            { isExpanded && (
                <div className={ styles.expandableBlock }>
                    {children}
                </div>
            )}
            <div className={ styles.toggle } onClick={() => handleClickToggle()}>
                { isExpanded ? effectiveCollapsePrompt : effectiveExpandPrompt }
            </div>
        </div>
    )
}

export default Expandable;