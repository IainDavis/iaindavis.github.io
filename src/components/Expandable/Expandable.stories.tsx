import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import Expandable from './Expandable';
import type { ExpandablePropsType } from './Expandable';
import Lorem from '/.storybook/assets/mdx/lorem_ipsum.mdx';

const render = (args: ExpandablePropsType) => (
    <>
      <h2>Lorem Ipsum...</h2>
      <Expandable {...args}>
        <div>
          <Lorem/>
        </div>
      </Expandable>
    </>
)

const meta: Meta<typeof Expandable> = {
  title: 'Custom Components/Expandable',
  component: Expandable,
  parameters: {
    layout: 'centered',
    docs: {
      source: {
        transform: (src: string) => src
          .replace(/^\s*$/g, '')
          .replace(/<MDXProvider>([\s\S]*?)<\/MDXProvider>/, '$1') 
          .replace(/__WEBPACK_DEFAULT_EXPORT__/g, 'Expandable')
          .replace(/MDXContent/g, "Lorem")
          .replace(/\n(  )/g, '\n') // repair indentation where the above adjustments have damaged it
      }
    }
  },
  argTypes: {
    startExpanded: {
      control: 'boolean',
      description: 'Should the section be expanded when the page first loads?'
    },
    expandPrompt: {
      control: 'text',
      description: 'The text to use when prompting the user to expand the region and reveal the hidden content'
    },
    collapsePrompt: {
      control: 'text',
      description: 'The text to use when prompting the user to collapse the region and hide the revealed content'
    }
  },
  render
};

export default meta;
type Story = StoryObj<typeof Expandable>;


// More on writing stories with args: https://storybook.js.org/docs/writing-stories/args
export const Default: Story = {
  args: {},
};

export const PreExpanded: Story = {
  args: {
    startExpanded: true
  },
};

export const CustomExpandCollapsePrompt: Story = {
  name: "Customizable Expand/Collapse Prompts",
  args: {
    expandPrompt: 'Click for more details (customized expand prompt)',
    collapsePrompt: 'Hide details (customized collapse prompt)',
  },
}

export const Nested: Story = {
  name: "Nested Expandable Sections",
  render: (args, context) => (
    <>
      <h1>Lorem Ipsum...</h1>
      <Expandable startExpanded>
        <div>
          <Lorem />
        </div>
        <Expandable startExpanded>
          Nested expanded content
        </Expandable>
      </Expandable>
    </>
  )
}

