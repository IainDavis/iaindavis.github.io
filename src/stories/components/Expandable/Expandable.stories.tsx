import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import { MDXProvider } from '@mdx-js/react';
import Expandable from '../../../components/Expandable';
import type { ExpandablePropsType } from '../../../components/Expandable';
import Lorem from '../../../../.storybook/assets/mdx/lorem_ipsum.mdx'

const render = (args: ExpandablePropsType) => (
  <MDXProvider>
    <Expandable {...args}>
      <Lorem/>
    </Expandable>
  </MDXProvider>
)

const meta = {
  title: 'Custom Components/Expandable',
  component: Expandable,
  parameters: {
    layout: 'centered',
    docs: {
      source: {
        transform: (src: string) => src
          .replace(/^\s*$/g, '')
          .replace(/<MDXProvider>([\s\S]*?)<\/MDXProvider>/, '$1') 
          .replace(/__WEBPACK_DEFAULT_EXPORT__/g, 'Expandable')
          .replace(/MDXContent/g, "Lorem")
          .replace(/\n(  )/g, '\n') // repair indentation where the above adjustments have damaged it
      }
    }
  },
  argTypes: {},
  args: {},
  render
} satisfies Meta<typeof Expandable>;

export default meta;
type Story = StoryObj<typeof meta>;


// More on writing stories with args: https://storybook.js.org/docs/writing-stories/args
export const Default: Story = {
  args: {},
};

export const PreExpanded: Story = {
  args: {
    startExpanded: true
  },
};

export const CustomExpandCollapsePrompt: Story = {
  name: "Customizable Expand/Collapse Prompts",
  args: {
    expandPrompt: 'Click for more details (customized expand prompt)',
    collapsePrompt: 'Hide details (customized collapse prompt)',
  },
}

export const Nested: Story = {
  name: "Nested Expandable Sections",
  render: (args, context) => (
    <MDXProvider>
      <Expandable startExpanded>
        <Lorem/>
        <Expandable startExpanded>
          Nested expanded content
        </Expandable>
      </Expandable>
    </MDXProvider>
  )
}

