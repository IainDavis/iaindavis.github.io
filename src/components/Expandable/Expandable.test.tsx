import React from 'react';

import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { composeStories } from '@storybook/react';

import * as stories from './Expandable.stories';

import Expandable from './Expandable';

const {
    Default: DefaultStory,
    PreExpanded: PreExpandedStory,
    CustomExpandCollapsePrompt: CustomExpandCollapsePromptStory,
    Nested: NestedStory
} = composeStories(stories);

describe ('Expandable Component', () => {
    describe('Basic tests', () => {
        
        // Test default collapsed state and expanded on click
        it('should render collapsed by default and expand when clicked', () => {
            render(<Expandable startExpanded={false} expandPrompt="Expand" />);
            expect(screen.getByText('Expand')).toBeInTheDocument();
            fireEvent.click(screen.getByText('Expand'));
            expect(screen.getByText('show less')).toBeInTheDocument();
        });

        // Test the component renders expanded when startExpanded is true
        it('should render expanded by default if startExpanded is true', () => {
            render(<Expandable startExpanded={true} collapsePrompt="Collapse" />);
            expect(screen.getByText('Collapse')).toBeInTheDocument();
        });

        // Test collapsing after being expanded
        it('should collapse when clicked after being expanded', () => {
            render(<Expandable startExpanded={true} collapsePrompt="Collapse" />);
            fireEvent.click(screen.getByText('Collapse'));
            expect(screen.getByText('show more...')).toBeInTheDocument();
        });

        // Handle missing or empty props gracefully
        it('should handle empty prompts gracefully', () => {
            render(<Expandable startExpanded={false} expandPrompt="" collapsePrompt="" />);
            fireEvent.click(screen.getByText('show more...'));
            expect(screen.getByText('show less')).toBeInTheDocument();
        });

        it('should handle null prompts gracefully', () => {
            render(<Expandable startExpanded={false} expandPrompt={null} collapsePrompt={null} />);
            fireEvent.click(screen.getByText('show more...'));
            expect(screen.getByText('show less')).toBeInTheDocument();
        });

        it('should handle undefined prompts gracefully', () => {
            render(<Expandable startExpanded={false} />);
            fireEvent.click(screen.getByText('show more...'));
            expect(screen.getByText('show less')).toBeInTheDocument();
        });

        // Test large content performance
        it('should perform well with large content when expanding/collapsing', () => {
            const largeContent = Array(1000).fill('Content').join(' ');
            render(<Expandable>{largeContent}</Expandable>);
            fireEvent.click(screen.getByText('show more...'));
            expect(screen.getByText('show less')).toBeInTheDocument();
        });

        // Check dynamic state changes (props change during lifecycle)
        it('should update when startExpanded prop is changed dynamically', () => {
            const { rerender } = render(<Expandable startExpanded={false} />);
            expect(screen.queryByText('show less')).toBeNull();

            rerender(<Expandable startExpanded={true} />);
            expect(screen.getByText('show less')).toBeInTheDocument();
        });

        // Visual regression - check for UI changes in different states
        it('should maintain visual consistency between expanded and collapsed states', () => {
            const { asFragment, rerender } = render(<Expandable startExpanded={false} />);
            expect(asFragment()).toMatchSnapshot();

            rerender(<Expandable startExpanded={true} />);
            expect(asFragment()).toMatchSnapshot();
        });

        // Check responsiveness/media queries behavior
        it('should respond correctly to different screen sizes', () => {
            global.innerWidth = 500;
            render(<Expandable />);
            expect(screen.getByText('show more...')).toBeInTheDocument();

            global.innerWidth = 1024;
            fireEvent.resize(window);
            expect(screen.getByText('show more...')).toBeInTheDocument();
        });
    });

    describe('Storybook Stories', () => {
        
        // Test for the "PreExpanded" story
        it('should render the "PreExpanded" story correctly', () => {
            render(<PreExpandedStory />);
            const content = screen.getByText(/exercitation ullamco laboris/i); // Replace with actual content from your story
            expect(content).toBeVisible();  // Content should be visible since it's pre-expanded
        });

        // Test for the "CustomExpandCollapsePrompt" story
        it('should render custom expand and collapse prompts in the "CustomExpandCollapsePrompt" story', () => {
            render(<CustomExpandCollapsePromptStory />);
            const expandButton = screen.getByText(/click for more details/i); // Custom expand prompt
            expect(expandButton).toBeInTheDocument();

            const content = screen.getByText(/exercitation ullamco laboris/i); // Replace with actual content from your story
            expect(content).not.toBeVisible();
            
            fireEvent.click(expandButton);  // Expand the content
            expect(content).toBeVisible();
            
            const collapseButton = screen.getByText(/hide details/i);  // Custom collapse prompt
            expect(collapseButton).toBeInTheDocument();
        });

        // Test for the "Default" story
        it('should render the "Default" story collapsed by default', () => {
            render(<DefaultStory />);
            const content = screen.queryByText(/exercitation ullamco laboris/i);  // Replace with actual content from your story
            expect(content).not.toBeVisible(); 
        });

        // Test for expanding the "Default" story
        it('should expand the "Default" story content when clicked', () => {
            render(<DefaultStory />);
            const expandButton = screen.getByText(/show more/i);  // Replace with the actual prompt from your story
            const content = screen.getByText(/exercitation ullamco laboris/i);  // Replace with actual content from your story

            expect(content).not.toBeVisible(); 
            fireEvent.click(expandButton);
            expect(content).toBeVisible(); 
        }); 

        it('should allow independent toggling of nested Expandable components', () => {
            render(<NestedStory/>);
            
            const [childButton, parentButton] = screen.getAllByRole('button', { name: /show less/})
            const outerContent = screen.getByText(/exercitation ullamco laboris/i);  // Replace with actual content from your story
            const innerContent = screen.getByText(/Nested expanded content/)
            
            // Initially, both sections are expanded
            expect(innerContent).toBeVisible();
            expect(outerContent).toBeVisible();
        
            // Collapse the parent
            fireEvent.click(parentButton);

            expect(outerContent).not.toBeVisible();
            expect(innerContent).not.toBeVisible();

            // (not realistic) collapse the child while the parent is collapsed
            fireEvent.click(childButton)

            // expand the parent
            fireEvent.click(parentButton);

            // ensure the child remains collapsed (independent behavior)
            expect(outerContent).toBeVisible();
            expect(innerContent).not.toBeVisible();
        
            // expand the child
            fireEvent.click(childButton);

            // both should be visible.
            expect(outerContent).toBeVisible();
            expect(innerContent).toBeVisible();
        });
    });

    describe('Accessibility Tests', () => {
        const expandPrompt = 'Show more';
        const collapsePrompt = 'Show less';

        const renderWithContent = ({ startExpanded } = { startExpanded: false }) => render(
            <Expandable expandPrompt={expandPrompt} collapsePrompt={collapsePrompt} startExpanded={startExpanded}>
                <p>Expandable content</p>
            </Expandable>
        )

        it('should render with initial collapsed state', () => {
            renderWithContent({ startExpanded: false })

            const button = screen.getByRole('button', { name: expandPrompt });
            const expandableSection = screen.getByTestId(/expandable-section/);

            expect(button).toHaveAttribute('aria-expanded', 'false');
            expect(expandableSection).toHaveAttribute('hidden');
        });

        it('should expand and focus on content when clicked', () => {
            renderWithContent();

            const button = screen.getByRole('button', { name: expandPrompt });
            const expandableSection = screen.getByTestId(/expandable-section/);

            fireEvent.click(button);

            expect(button).toHaveAttribute('aria-expanded', 'true');
            expect(expandableSection).not.toHaveAttribute('hidden');
            expect(expandableSection).toHaveFocus();
        });

        it('should collapse and move focus back to the button', () => {
            renderWithContent({ startExpanded: true});

            const button = screen.getByRole('button', { name: collapsePrompt });
            const expandableSection = screen.getByTestId(/expandable-section/);

            fireEvent.click(button);

            expect(button).toHaveAttribute('aria-expanded', 'false');
            expect(expandableSection).toHaveAttribute('hidden');
            expect(button).toHaveFocus();
        });

        it('should start expanded if startExpanded is true', () => {
            renderWithContent({ startExpanded: true })

            const button = screen.getByRole('button', { name: collapsePrompt });
            const expandableSection = screen.getByTestId(/expandable-section/);

            expect(button).toHaveAttribute('aria-expanded', 'true');
            expect(expandableSection).not.toHaveAttribute('hidden');
        });

        describe('When Expandables are nested', () => {

            const renderNestedExpandable = () => render(
                <Expandable expandPrompt="Show more" collapsePrompt="Show less" >
                    <p>Outer expandable content</p>
                    <Expandable expandPrompt="Show nested" collapsePrompt="Hide nested" >
                        <p>Nested expandable content</p>
                    </Expandable>
                </Expandable>
            );

            it('should set aria-expanded correctly for parent and child independently', async () => {
                renderNestedExpandable();
        
                const parentButton = screen.getByText(/show more/i, { selector: "button"});
                const childButton = screen.getByText(/show nested/i, { selector: "button"});

                // Initial state
                expect(parentButton).toHaveAttribute('aria-expanded', 'false');
                expect(childButton).toHaveAttribute('aria-expanded', 'false');
        
                // Expand parent
                fireEvent.click(parentButton);
                expect(parentButton).toHaveAttribute('aria-expanded', 'true');
        
                // Child remains collapsed
                expect(childButton).toHaveAttribute('aria-expanded', 'false');
        
                // Expand child
                fireEvent.click(childButton);
                expect(childButton).toHaveAttribute('aria-expanded', 'true');
        
                // Both should now be expanded
                expect(parentButton).toHaveAttribute('aria-expanded', 'true');
                expect(childButton).toHaveAttribute('aria-expanded', 'true');
            });
        
            it('should manage focus correctly between parent and child expandables', async () => {
                renderNestedExpandable();
        
                const parentButton = screen.getByText(/show more/i, { selector: "button"});
                const childButton = screen.getByText(/show nested/i, { selector: "button"});

                // focus gets applied to the container, not to the actual content.
                // actual content might well start with an unfocusable element like <p>
                const parentContent = screen.getByText(/outer expandable content/i).closest('div');
                const childContent = screen.getByText(/nested expandable content/i).closest('div');
        
                // Expand parent and check focus on content
                fireEvent.click(parentButton);
                await waitFor(() => expect(parentContent).toHaveFocus());

                // expect(parentContent).toHaveFocus();
        
                // Expand child and check focus on nested content
                fireEvent.click(childButton);
                expect(childContent).toHaveFocus();
        
                // Collapse parent, focus should move back to the button
                fireEvent.click(parentButton);
                expect(parentButton).toHaveFocus();
            });
        
            it('should be keyboard navigable for both parent and child', () => {
                renderNestedExpandable();
        
                const parentButton = screen.getByText(/show more/i, { selector: "button"});
                const childButton = screen.getByText(/show nested/i, { selector: "button"});
        
                // Simulate keyboard events to expand/collapse
                fireEvent.keyDown(parentButton, { key: 'Enter' });
                expect(parentButton).toHaveAttribute('aria-expanded', 'true');
        
                // Now simulate for the child component
                fireEvent.keyDown(childButton, { key: 'Enter' });
                expect(childButton).toHaveAttribute('aria-expanded', 'true');
        
                // Collapse both with space key
                fireEvent.keyDown(childButton, { key: ' ' });
                expect(childButton).toHaveAttribute('aria-expanded', 'false');
        
                fireEvent.keyDown(parentButton, { key: ' ' });
                expect(parentButton).toHaveAttribute('aria-expanded', 'false');
            });
        });
    });

})




        