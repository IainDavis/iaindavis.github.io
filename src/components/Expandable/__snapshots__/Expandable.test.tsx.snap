// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Expandable Component > Basic tests > should maintain visual consistency between expanded and collapsed states 1`] = `
<DocumentFragment>
  <div
    class="_container_c6a052"
  >
    <div
      class="_expandableBlock_c6a052"
      data-testid="expandable-section-mocked-id"
      hidden=""
      id="expandable-section-mocked-id"
      tabindex="1"
    />
    <button
      aria-controls="expandable-section-toggle-mocked-id"
      aria-expanded="false"
      class="_toggle_c6a052"
      tabindex="0"
    >
      show more...
    </button>
  </div>
</DocumentFragment>
`;

exports[`Expandable Component > Basic tests > should maintain visual consistency between expanded and collapsed states 2`] = `
<DocumentFragment>
  <div
    class="_container_c6a052"
  >
    <div
      class="_expandableBlock_c6a052"
      data-testid="expandable-section-mocked-id"
      id="expandable-section-mocked-id"
      tabindex="0"
    />
    <button
      aria-controls="expandable-section-toggle-mocked-id"
      aria-expanded="true"
      class="_toggle_c6a052"
      tabindex="0"
    >
      show less
    </button>
  </div>
</DocumentFragment>
`;
